ios/tunnel/codec.go

```diff
@@ -0,0 +1,218 @@
+	Decode(e map[string]interface{}) error
+}
+
+type pairingData struct {
+	data            []byte
+	kind            string
+	sendingHost     string
+	startNewSession bool
+}
+
+func (p *pairingData) Encode() map[string]interface{} {
+	return map[string]interface{}{
+		"pairingData": map[string]interface{}{
+			"_0": map[string]interface{}{
+				"data":            p.data,
+				"kind":            p.kind,
+				"sendingHost":     p.sendingHost,
+				"startNewSession": p.startNewSession,
+			},
+		},
+	}
+}
+
+func (p *pairingData) Decode(e map[string]interface{}) error {
+	pd, err := getChildMap(e, "pairingData", "_0")
+	if err != nil {
+		return err
+	}
+	if data, ok := pd["data"].([]byte); ok {
+		p.data = data
+	}
+	if kind, ok := pd["kind"].(string); ok {
+		p.kind = kind
+	}
+	if startNewSession, ok := pd["startNewSession"].(bool); ok {
+		p.startNewSession = startNewSession
+	}
+	if sendingHost, ok := pd["sendingHost"].(string); ok {
+		p.sendingHost = sendingHost
+	}
+	return nil
+}
+
+type pairVerifyFailed struct {
+}
+
+func (p pairVerifyFailed) Encode() map[string]interface{} {
+	return map[string]interface{}{
+		"pairVerifyFailed": map[string]interface{}{},
+	}
+}
+
+func (p pairVerifyFailed) Decode(e map[string]interface{}) error {
+	return nil
+}
+
+func getChildMap(m map[string]interface{}, keys ...string) (map[string]interface{}, error) {
+	if len(keys) == 0 {
+		return m, nil
+	}
+	if c, ok := m[keys[0]].(map[string]interface{}); ok {
+		return getChildMap(c, keys[1:]...)
+	} else {
+		return nil, fmt.Errorf("something went wrong")
+	}
+}
+
+type xpcConn interface {
+	Send(data map[string]interface{}, flags ...uint32) error
+	ReceiveOnClientServerStream() (map[string]interface{}, error)
+}
+
+type controlChannelReadWriter struct {
+	seqNr uint64
+	conn  xpcConn
+}
+
+func newControlChannelReadWriter(conn xpcConn) *controlChannelReadWriter {
+	return &controlChannelReadWriter{
+		seqNr: 1,
+		conn:  conn,
+	}
+}
+
+func (c *controlChannelReadWriter) writeEventRaw(e map[string]interface{}) error {
+	panic(nil)
+}
+
+func (c *controlChannelReadWriter) writeEvent(e eventCodec) error {
+	encoded := map[string]interface{}{
+		"plain": map[string]interface{}{
+			"_0": map[string]interface{}{
+				"event": map[string]interface{}{
+					"_0": e.Encode(),
+				},
+			},
+		},
+	}
+	return c.write(encoded)
+}
+
+func (c *controlChannelReadWriter) readEvent(e eventCodec) error {
+	m, err := c.read()
+	if err != nil {
+		return err
+	}
+	event, err := getChildMap(m, "plain", "_0", "event", "_0")
+	if err != nil {
+		return err
+	}
+	return e.Decode(event)
+}
+
+func (c *controlChannelReadWriter) writeRequest(req map[string]interface{}) error {
+	return c.write(map[string]interface{}{
+		"plain": map[string]interface{}{
+			"_0": map[string]interface{}{
+				"request": map[string]interface{}{
+					"_0": req,
+				},
+			},
+		},
+	})
+}
+
+func (c *controlChannelReadWriter) write(message map[string]interface{}) error {
```

Comment:
would be nice to explain a bit what these functions do and what the difference between write and write Request ist